src/chai/ArrayManager.cpp

```diff
@@ -209,17 +223,27 @@ void ArrayManager::move(PointerRecord* record, ExecutionSpace space)
     return;
   }
 
+  callback(record, ACTION_CAPTURED, space);
+
+  if (space == record->m_last_space) {
+    return;
+  }
+
 #if defined(CHAI_ENABLE_UM)
   if (record->m_last_space == UM) {
     return;
   }
 #endif
 
-  callback(record, ACTION_CAPTURED, space);
-
-  if (space == record->m_last_space) {
+#if defined(CHAI_ENABLE_PINNED)
+  if (record->m_last_space == PINNED) {
+    if (space == CPU && m_need_sync_for_pinned) {
+      m_need_sync_for_pinned = false;
+      cudaDeviceSynchronize();
```

Comment:
This should be if-defd for HIP/CUDA, maybe hide it in a `chai::synchronize()` call or something.

Reply:
I've added a chai::synchronize that should handle hip and cuda.